parse-zoneinfo/src/line.rs

```diff
@@ -628,12 +623,64 @@ pub enum TimeType {
     UTC,
 }
 
+impl TimeType {
+    fn from_char(c: char) -> Option<Self> {
+        Some(match c {
+            'w' => Self::Wall,
+            's' => Self::Standard,
+            'u' | 'g' | 'z' => Self::UTC,
+            _ => return None,
+        })
+    }
+}
+
 #[derive(PartialEq, Debug, Copy, Clone)]
 pub struct TimeSpecAndType(pub TimeSpec, pub TimeType);
 
-impl TimeSpec {
-    pub fn with_type(self, timetype: TimeType) -> TimeSpecAndType {
-        TimeSpecAndType(self, timetype)
+impl FromStr for TimeSpecAndType {
+    type Err = Error;
+
+    fn from_str(input: &str) -> Result<Self, Error> {
```

Comment:
Can you please split this method over the `TimeSpec` and `TimeSpecAndType` types? I am an not sure yet if anything but wall times is allowed zone lines, and if the existing code took a shortcut there that we want to fix.